illumos-utils/src/chronyd.rs

```diff
@@ -0,0 +1,24 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! Utilities for interacting with chronyd.
+
+use crate::zone::CHRONYD;
+use crate::{execute, ExecutionError, PFEXEC};
+
+/// Wraps commands for interacting with chronyd.
+pub struct Chronyd {}
+
+#[cfg_attr(any(test, feature = "testing"), mockall::automock)]
+impl Chronyd {
+    pub fn start_daemon(file: &str) -> Result<(), ExecutionError> {
+        let mut cmd = std::process::Command::new(PFEXEC);
+        // TODO: This doesn't seem to be working. I think `execute()`
+        // doesn't like the "&", and it immediately exits after running.
```

Comment:
The `&` is a shell construct, but here you're not using a shell, you're just making arguments to pass to an **execve(2)** call ultimately.  It seems like the original method script is doing this because we're using the `-d` flag for chrony, which has the positive effect of using stdout for logging (which we want) but the negative effect of then not daemonising itself.

What I would do here is use the **exec()** method on Rust's **Command** object, to _replace_ this process with chrony.  I would then change the service **duration** to **child** for this service manifest, so that SMF expects we'll keep running forever without detaching.

Also no pfexec should be necessary, we can arrange the correct set of credentials and privileges in the SMF service's method context.

Reply:
Another option, particularly since I think it makes sense to keep the method script for the stop and refresh methods so they can properly manage the contract using the framework from `/lib/svc/share/smf_include.sh`, is to make a new chrony setup service that does the configuration, and make the existing chrony service dependant on that. That could all still be done within the one manifest - an example of a manifest with multiple service instances is https://github.com/omniosorg/cloud-init/blob/illumos/smf/cloud-init.xml; in this case all of the instances share the stop and refresh methods, but that isn't mandatory.

This would possibly be better from a privilege separation perspective too. The current service manifest starts the service as the root user, but without full root privileges - it actually starts by applying the `basic` set, removing some, and then adding a few which are required for the NTP daemon. These aren't needed for the setup script which effectively just needs to write to a couple of files - for example it definitely doesn't need/shouldn't have `sys_time`.

```xml
  <exec_method type="method" name="start" exec="/var/svc/method/svc-site-ntp %m"
    timeout_seconds="60">
  <method_context security_flags="aslr">
  <method_credential user="root" group="root" privileges="basic,!file_link_any,!proc_info,!proc_session,file_chown_self,file_dac_search,file_dac_write,net_privaddr,proc_lock_memory,proc_priocntl,proc_setid,sys_time" />
  </method_context>
  </exec_method>
```

Reply:
Thanks both! :bowing_woman: 

> Another option, particularly since I think it makes sense to keep the method script for the stop and refresh methods so they can properly manage the contract using the framework from /lib/svc/share/smf_include.sh, is to make a new chrony setup service that does the configuration, and make the existing chrony service dependant on that.

Hmm, this makes sense. I'm assuming that I will the be able to safely remove `generate_config_file` from the refresh method and it will automatically generate the config file as the ntp service now depends on the chrony setup service?

```
refresh)
	generate_config_file && stop_daemon
	# SMF will restart the service since the contract is now empty.
	;;
```

Reply:
Moving configuration file generation to a separate **transient** duration service makes sense to me.  I am unclear on the value of the refresh method here, such as it is.  If you're just going to kill the program, just restart the service?  Then you probably don't need any of the shell at all, I suspect you could just execute chrony directly with the **child** duration.

Reply:
Also I looked at the cloud-init service bundle and it's probably not exactly the example I would follow here.  The configuration generation and the actual time daemon are notionally different _services_, rather than _instances_ of the same service.  If you're going to co-locate them in a bundle they should be separate services at least.

Reply:
Thanks @jclulow! What I'm understanding from @citrus-it's comment, is that we want to be able to keep the `stop_daemon` function to properly manage the contract. So we'd just keep the tiniest bit of the script.

```
function stop_daemon {
	echo "* Stopping daemon"
	smf_kill_contract $contract TERM 1 30
	rc=$?
	((rc == 1)) && fatal "Invalid contract in $action"
	# It's possible that the contract did not empty with SIGTERM; move on
	# to KILL.
	((rc == 2)) && smf_kill_contract $contract KILL 1
}
```

Please someone correct me if I'm wrong though!

Reply:
That's the part that I don't understand!  Stopping the processes in the contract is the first thing that disable or restart does.

Reply:
I'll try and find the background for this, it came from OmniOS. I've seen this technique done for other services like `sendmail` that fork off separate handlers for incoming connections, but I don't know why the standard startd restarter's behaviour is undesirable, unless it kills things too quickly.

With chrony, we need a soft-ish shutdown in general so that it can write its `driftfile` and what have you, and it runs as two processes, one of which is privileged, but I don't know the background.

Since this is in there, and in almost no other service delivered by omnios, I assume it was found to be load bearing in some way!

Reply:
Ahhhh!!! gotcha. @citrus-it, what do you think?

Reply:
thanks @citrus-it !

Reply:
Ok, in the end it was not from OmniOS - there there is just a simple kill. Same in OpenIndiana.
I don't remember why I added this in the service delivered from omicron, I presumably came across an instance where using kill as the stop method put the service into maintenance with an escaped process. It's usually to work around the race between killing the contract and the contract member forking - such as happens frequently with sendmail. I don't think `chrony` forks much once it's up and running though, especially as we don't configure it to invoke notification scripts. In fact we remove the ability to fork/exec at all from one of its processes.

Long story short, we can probably lose the method script entirely and just have:

```
        <exec_method type="method"
                     name="start"
                     exec="/usr/sbin/chronyd -d &amp;"
                     timeout_seconds="600">
...
        <exec_method type="method"
                     name="stop"
                     exec=":kill"
                     timeout_seconds="60" />
```

and a dependency on the new `chrony-configure` service that is responsible for setting things up before chrony starts.
The refresh method was originally for forcing a regeneration of config files and restarting the daemon, after the control plane had updated the various `config/` properties after a change, and issued a refresh.

This isn't wired up today, but it would still be good to arrange things so that this can work as expected, because we will one day support operators changing the upstream NTP servers etc.

Reply:
Thanks for the detailed explanation @citrus-it :bowing_woman:! It's interesting to learn more about Illumos.

> Long story short, we can probably lose the method script entirely and just have:

sweet! I'll adapt the NTP service and corresponding new chrony-setup service then.

Reply:
Thanks for the help both! I've set the PR as ready to review :)